src/compiler/tsbuildPublic.ts

```diff
@@ -1459,80 +1506,182 @@ namespace ts {
-                if (!force && !missingOutputFileName) {
-                    // If the upstream project's newest file is older than our oldest output, we
-                    // can't be out of date because of it
-                    if (refStatus.newestInputFileTime && refStatus.newestInputFileTime <= oldestOutputFileTime) {
-                        continue;
-                    }
+                if (!force) (referenceStatuses ||= []).push({ ref, refStatus, resolvedRefPath, resolvedConfig });
+            }
+        }
+        if (force) return { type: UpToDateStatusType.ForceBuild };
 
-                    // If the upstream project has only change .d.ts files, and we've built
-                    // *after* those files, then we're "psuedo up to date" and eligible for a fast rebuild
-                    if (refStatus.newestDeclarationFileContentChangedTime && refStatus.newestDeclarationFileContentChangedTime <= oldestOutputFileTime) {
-                        pseudoUpToDate = true;
-                        upstreamChangedProject = ref.path;
-                        continue;
-                    }
+        // Check buildinfo first
+        const { host } = state;
+        const buildInfoPath = getTsBuildInfoEmitOutputFilePath(project.options);
+        let oldestOutputFileName = "(none)";
+        let oldestOutputFileTime = maximumDate;
+        let buildInfoTime: Date | undefined;
+        let newestDeclarationFileContentChangedTime;
+        if (buildInfoPath) {
+            const buildInfoCacheEntry = getBuildInfoCacheEntry(state, buildInfoPath, resolvedPath);
+            buildInfoTime = buildInfoCacheEntry?.modifiedTime || ts.getModifiedTime(host, buildInfoPath);
+            if (buildInfoTime === missingFileModifiedTime) {
+                if (!buildInfoCacheEntry) {
+                    state.buildInfoCache.set(resolvedPath, {
+                        path: toPath(state, buildInfoPath),
+                        buildInfo: false,
+                        modifiedTime: buildInfoTime
+                    });
+                }
+                return {
+                    type: UpToDateStatusType.OutputMissing,
+                    missingOutputFileName: buildInfoPath
+                };
+            }
+
+            const buildInfo = Debug.checkDefined(getBuildInfo(state, buildInfoPath, resolvedPath, buildInfoTime));
+            if ((buildInfo.bundle || buildInfo.program) && buildInfo.version !== version) {
+                return {
+                    type: UpToDateStatusType.TsVersionOutputOfDate,
+                    version: buildInfo.version
+                };
+            }
 
-                    // We have an output older than an upstream output - we are out of date
-                    Debug.assert(oldestOutputFileName !== undefined, "Should have an oldest output filename here");
+            if (buildInfo.program) {
+                if ((buildInfo.program as ProgramMultiFileEmitBuildInfo).changeFileSet?.length ||

```

Comment:
If build info has changeSet or files pending emit (only in emit) scenarios, we need build. We dont need to check other output time stamps as buildInfo is the truth.